lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -690,28 +690,41 @@ std::pair<uint64_t, bool> FIRRTLBaseType::rootChildFieldID(uint64_t fieldID,
 /// canonicalizes flips in bundles, so only passive types can be compared here.
 static bool areBundleElementsEquivalent(BundleType::BundleElement destElement,
                                         BundleType::BundleElement srcElement,
+                                        bool destOuterTypeIsConst,
                                         bool srcOuterTypeIsConst,
                                         bool requiresSameWidth) {
   if (destElement.name != srcElement.name)
     return false;
   if (destElement.isFlip != srcElement.isFlip)
     return false;
 
+  if (destElement.isFlip) {
+    std::swap(destElement, srcElement);
+    std::swap(destOuterTypeIsConst, srcOuterTypeIsConst);
+  }
+
   return areTypesEquivalent(destElement.type, srcElement.type,
-                            srcOuterTypeIsConst, requiresSameWidth);
+                            destOuterTypeIsConst, srcOuterTypeIsConst,
+                            requiresSameWidth);
 }
 
 /// Returns whether the two types are equivalent.  This implements the exact
 /// definition of type equivalence in the FIRRTL spec.  If the types being
 /// compared have any outer flips that encode FIRRTL module directions (input or
 /// output), these should be stripped before using this method.
 bool firrtl::areTypesEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
+                                bool destOuterTypeIsConst,
                                 bool srcOuterTypeIsConst,
                                 bool requireSameWidths) {
 
-  // If the types are trivially equal, return true.
-  if (destFType == srcFType)
-    return true;
+  // If the types are trivially equal, return true as long as outer constness is
+  // compatible.
+  if (destFType == srcFType) {
+    if (destOuterTypeIsConst == srcOuterTypeIsConst)
+      return true;
+    if (destFType.isGround())
+      return !destOuterTypeIsConst || srcOuterTypeIsConst;
+  }
```

Comment:
Hmm maybe we should just get rid of the early break.  I think I added this recently, and introduced the bug of ignoring the outer const type.  If we ever push const to the leaves or something like that, then maybe we can introduce the early break again, wdyt?

Also I suggest we add a test case for this because I keep thinking we can add this trivial early break.  Its now covered by the flip test cases you added, but the most basic test might be:
```mlir
/// Const types cannot be connected to non-const types.

firrtl.circuit "test" {
firrtl.module @test(in %in   : !firrtl.const.bundle<a: uint<1>>,
                    out %out : !firrtl.bundle<a: uint<1>>) {
  // expected-error @+1 {{type mismatch}}
  firrtl.connect %out, %in : !firrtl.bundle<a: uint<1>>, !firrtl.const.bundle<a: uint<1>>
}
}
```

Reply:
Yeah, the outer constness gets kind of unwieldy, I like your suggestion of keeping things in one place for ground types.
Removed, and added test case (I assume you accidentally wrote the valid version).